File: src/main/java/editor/filter/leaf/options/multi/TagsFilter.java
package editor.filter.leaf.options.multi;

import java.util.HashSet;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

import editor.database.attributes.CardAttribute;
import editor.database.card.Card;
import editor.filter.leaf.FilterLeaf;

import scala.jdk.javaapi.CollectionConverters;

/**
 * This class represents a filter that filters cards by user-controlled tags.
 *
 * @author Alec Roelke
 */
public class TagsFilter extends MultiOptionsFilter<String>
{
    /**
     * Create a new TagsFilter.
     */
    public TagsFilter()
    {
        super(CardAttribute.TAGS, (c) -> {
            if (Card.TAGS().contains(c))
                return CollectionConverters.asJava(Card.TAGS().apply(c));
            else
                return new HashSet<String>();
        });
    }

    @Override
    protected String convertFromString(String str)
    {
        return str;
    }

    @Override
    protected FilterLeaf<String> copyLeaf()
    {
        TagsFilter filter = (TagsFilter)CardAttribute.createFilter(CardAttribute.TAGS);
        filter.contain = contain;
        filter.selected = new HashSet<>(selected);
        return filter;
    }

    @Override
    protected JsonElement convertToJson(String item)
    {
        return new JsonPrimitive(item);
    }

    @Override
    protected String convertFromJson(JsonElement item)
    {
        return item.getAsString();
    }
}

File: src/main/scala/editor/filter/leaf/options/multi/TagsFilter.scala
package editor.filter.leaf.options.multi

import editor.database.attributes.CardAttribute
import editor.database.card.Card
import editor.filter.FaceSearchOptions
import editor.util.Containment

/**
 * Filter that groups cards by user-defined tags.
 * @author Alec Roelke
 */
final case class TagsFilter(contain: Containment = Containment.AnyOf, selected: Set[String] = Set.empty) extends MultiOptionsFilter[String, TagsFilter] {
  override def faces = FaceSearchOptions.ANY
  override def attribute = CardAttribute.Tags
  override val unified = true
  override def values = Card.tags.get(_).map(_.toSet).getOrElse(Set.empty)
  override def copy(faces: FaceSearchOptions, contain: Containment, selected: Set[String]) = copy(contain = contain, selected = selected)
  override def copyFaces(faces: FaceSearchOptions) = this
}